src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -294,22 +305,22 @@ SafeToPushdownUnionSubquery(PlannerRestrictionContext *plannerRestrictionContext
 		}
 		else
 		{
-			partitionKeyIndex =
-				RelationRestrictionPartitionKeyIndex(relationRestriction);
+			varToBeAdded =
+				RelationRestrictionPartitionKeyIndex(originalQuery, relationRestriction,
+													 &partitionKeyIndex);
 
 			/* union does not have partition key in the target list */
 			if (partitionKeyIndex == 0)
 			{
 				continue;
 			}
 
-			targetEntryToAdd = list_nth(targetList, partitionKeyIndex - 1);
-			if (!IsA(targetEntryToAdd->expr, Var))
-			{
-				continue;
-			}
-
-			varToBeAdded = (Var *) targetEntryToAdd->expr;
+			/*
+			 * we update the varno because we use the original parse tree for finding the
+			 * var. However the rest of the code relies on a query tree that might be different
+			 * than the original parse tree because of postgres optimizations.
+			 */
+			varToBeAdded->varno = relationRestriction->index;
```

Comment:
we already return the var from `relationPlannerParseQuery`. Do we really need to set this if we keep the code as-is?

Reply:
> relationPlannerParseQuery

as we talked, that's a naming issue. we should keep this